src/transformers/models/llava/modeling_llava.py

```diff
@@ -409,25 +477,79 @@ def forward(
         )
 
         if inputs_embeds is None:
-            # 1. Extra the input embeddings
+            # 1. Extract the input embeddings
             inputs_embeds = self.get_input_embeddings()(input_ids)
 
             # 2. Merge text and images
             if pixel_values is not None and input_ids.shape[1] != 1:
-                image_outputs = self.vision_tower(pixel_values, output_hidden_states=True)
-                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
-                selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
-
-                if vision_feature_select_strategy == "default":
-                    selected_image_feature = selected_image_feature[:, 1:]
-                elif vision_feature_select_strategy == "full":
-                    selected_image_feature = selected_image_feature
+                if pixel_values.ndim == 5:
+                    concat_images = torch.cat(list(pixel_values), dim=0)
+                    image_features = self.vision_tower(concat_images, output_hidden_states=True)
+
+                    selected_image_feature = image_features.hidden_states[vision_feature_layer]
+
+                    if vision_feature_select_strategy == "default":
+                        selected_image_feature = selected_image_feature[:, 1:]
+                    elif vision_feature_select_strategy == "full":
+                        selected_image_feature = selected_image_feature
+                    else:
+                        raise ValueError(
+                            f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}"
+                        )
+                    image_features = self.multi_modal_projector(selected_image_feature)
+
+                    split_sizes = [image.shape[0] for image in pixel_values]
+                    image_features = torch.split(image_features, split_sizes, dim=0)
+
+                    # NOTE we only support multimodal_patch_merge_type == "spatial_unpad"
```

Comment:
This should be asserted somewhere and added to a docstring - most users won't read the code here

Reply:
This doesn't need to be asserted since `multimodal_patch_merge_type` isn't a configurable parameter, I've removed it since this is the only patch merging type that the authors eventually used. Hence the comment which I added